src/js/notes.md

```diff
@@ -0,0 +1,179 @@
+# Gleam to JS
+
+Working through the Gleam book to see how the language concepts translate to JS.
+
+### Comments, Strings, Bools, Ints and Floats
+These should all be trivial to map.
+
+*Integers in JS are not "efficient" but all the Gleam operators should work appropriately and return more integers.*
+
+From Gleam book and translated JS
+```rs
+1 + 1 // => 2
+5 - 1 // => 4
+5 / 2 // => 2
+3 * 3 // => 9
+5 % 2 // => 1
+```
+
+```js
+1 + 1 // => 2
+5 - 1 // => 4
+parseInt(5 / 2) // => 2
+3 * 3 // => 9
+5 % 2 // => 1
+```
+
+JS has no way to check if something is a float or integer reliably.
+`Number.isInteger(1.0) // => false`
+
+### Let bindings
+JS same.
+
+### Lists
+
+*JS lists are Arrays, so they will not have the same performance charachteristics.*
+However they match very neatly for syntax
+
+```rs
+let [a, ..b] = [1, 2, 3]
+let c = [4, ..b]
+let d = [4, ..b]
+```
+```js
+let [a, ...b] = [1, 2, 3]
+if (a === undefined) throw "match error"
+let c = [4, ...b]
+let d = [4, ...b]
+```
+
+Note the first line is a match error if list length is less than 1.
+
+### Tuple
+
+Probably best covered by a JS array, which can be heterogeneous.
+
+**Is it a problem that the dynamic library can't tell the difference between tuple and list in this case?**
+
+### Case Expressions
+
+JS `switch` statement uses strict equality and is not expressive enough.
```

Comment:
This gets even more complex when we consider nested patterns. There's some papers out there on implementing this efficiently, but a naive version with linear rather than logarithmic dispatch time would be OK. That's the norm in JavaScript anyway.

We can also look into what bucklescript does here.